src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/HotReload/ProjectHotReloadSessionManager.cs

```diff
@@ -201,24 +202,54 @@ async Task ActivateSessionInternalAsync()
         {
             if (_pendingSessionState is { Session: IProjectHotReloadSession session })
             {
-                Process process = Process.GetProcessById((int)vsDebugTargetProcessInfo.dwProcessId);
-                _pendingSessionState.LaunchedProcess = launchedProcess;
-                _pendingSessionState.Process = process;
+                Process? process = null;
+                try
+                {
+                    _pendingSessionState.LaunchedProcess = launchedProcess;
+                    process = Process.GetProcessById((int)vsDebugTargetProcessInfo.dwProcessId);
+                    _pendingSessionState.Process = process;
+                }
+                catch (ArgumentException)
+                {
+                    // process might have been exited in some cases.
+                    // in that case, we early return without starting hotreload session
+                    // one way to mimic this is to hit control + C as fast as you can once hit F5/Control + F5
+                    _pendingSessionState = null;
+                    return;
+                }
+
+                CancellationTokenSource cts = new CancellationTokenSource();
+                process.Exited += (sender, e) =>
+                {
+                    DebugTrace("Process exited");
+                    cts.Cancel();
+                };
 
                 if (!process.HasExited)
                 {
-                    process.Exited += (sender, e) =>
-                    {
-                        _threadingService.ExecuteSynchronously(() => session.StopSessionAsync(CancellationToken.None));
-                    };
                     process.EnableRaisingEvents = true;
-
-                    await _pendingSessionState.Session.StartSessionAsync(CancellationToken.None);
-                    lock (_activeSessionStates)
+                    await _pendingSessionState.Session.StartSessionAsync(cts.Token);
+                    if (cts.IsCancellationRequested)
```

Comment:
There's a race here where if the cancellation token is canceled after the sesssion is started but before the below registration, stop session is never called.